pixeldata/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Missing required element"))]
+    MissingRequiredField { source: dicom_object::Error },
+
+    #[snafu(display("Could not cast element"))]
+    CastValueError {
+        source: dicom_core::value::CastValueError,
+    },
+
+    #[snafu(display("Non supported GDCM PhotometricInterpretation: {}", pi))]
+    GDCMNonSupportedPI {
+        source: gdcm_rs::InvalidGDCMPI,
+        pi: String,
+    },
+
+    #[snafu(display("Non supported GDCM TransferSyntax: {}", ts))]
+    GDCMNonSupportedTS {
+        source: gdcm_rs::InvalidGDCMTS,
+        ts: String,
+    },
+
+    #[snafu(display("Invalid PixelData"))]
+    InvalidPixelData,
+
+    #[snafu(display("Invalid PixelRepresentation, must be 0 or 1"))]
+    InvalidPixelRepresentation,
+
+    #[snafu(display("Invalid BitsAllocated, must be 8 or 16"))]
+    InvalidBitsAllocated,
+
+    #[snafu(display("Unsupported PhotometricInterpretation {}", pi))]
+    UnsupportedPhotometricInterpretation { pi: String },
+
+    #[snafu(display("Unsupported SamplesPerPixel {}", spp))]
+    UnsupportedSamplesPerPixel { spp: u16 },
+
+    #[snafu(display("Unsupported TransferSyntax {}", ts))]
+    UnsupportedTransferSyntax { ts: String },
+
+    #[snafu(display("Multi-frame dicoms are not supported"))]
+    UnsupportedMultiFrame,
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Decoded pixel data
+/// NOTE: Is it possible to reuse dicom::PixelData ?
+pub struct DecodedPixelData {
+    pub data: Vec<u8>,
+    pub rows: u32,
+    pub cols: u32,
+    pub endianness: Endianness,
+    pub photometric_interpretation: String,
+    pub samples_per_pixel: u16,
+    pub bits_allocated: u16,
+    pub bits_stored: u16,
+    pub high_bit: u16,
+    pub pixel_representation: u16,
+}
+
+impl DecodedPixelData {
+    /// Convert decoded pixel data into a DynamicImage
+    pub fn to_dynamic_image(&self) -> Result<DynamicImage> {
+        if self.photometric_interpretation != "MONOCHROME2" {
+            UnsupportedPhotometricInterpretation {
+                pi: self.photometric_interpretation.clone(),
+            }
+            .fail()?
+        }
+
+        match self.bits_allocated {
+            8 => match self.samples_per_pixel {
+                1 => {
+                    // Single grayscale channel
+                    let image_buffer: ImageBuffer<Luma<u8>, Vec<u8>> =
+                        ImageBuffer::from_raw(self.cols, self.rows, self.data.to_owned()).unwrap();
+                    return Ok(DynamicImage::from(DynamicImage::ImageLuma8(image_buffer)));
+                }
+                _ => {
+                    // RGB, YBR, etc. color space
+                    UnsupportedSamplesPerPixel {
+                        spp: self.samples_per_pixel,
+                    }
+                    .fail()?
+                }
+            },
+            16 => {
+                let mut dest = vec![0; self.data.len() / 2];
```

Comment:
The vector created here is not used when `pixel_representation` is 1, so this other vector could be moved into the "unsigned" scope.

Reply:
It is used at the end because we normalize everything to u16 vectors. Otherwise we can't create an `Image` from it becauses it can only handle `u8` and `u16` vectors.